File: binarySearchTree/bst.py
class BST:
    def __init__(self,key):
        self.key = key
        self.lchild = None
        self.rchild = None
        
    def insert(self,data):
        if self.key is None:
            self.key = data
            return
        if (data<self.key) :
            if self.lchild :
                self.lchild.insert(data)
            else:
                self.lchild = BST(data)
        else:
            if self.rchild :
                self.rchild.insert(data)
            else:
                self.rchild = BST(data)
            

    def search(self,data):
        if self.key == data:
            print("present")
            return
        if data < self.key:
            if self.lchild:
                self.lchild.search(data)
            else:
                print("absent")
        else:
            if self.rchild:
                self.rchild.search(data)
            else:
                print("absent")

    def preorder(self):
        print(self.key)
        if self.lchild:
            self.lchild.preorder()
        if self.rchild:
            self.rchild.preorder()
        
    def inorder(self):
        if self.lchild:
            self.lchild.inorder()
        print(self.key)
        if self.rchild:
            self.rchild.inorder()
        
    def postorder(self):
        if self.lchild:
            self.lchild.inorder()
        if self.rchild:
            self.rchild.inorder()
        print(self.key)


    def delete(self,data):
        if not self.key:
            print("empty tree")
            return
        if data < self.key:
            if self.lchild:
                self.lchild=self.lchild.delete(data)
            else:
                print("not present")
        elif data > self.key:
            if self.rchild:
                self.rchild=self.rchild.delete(data)
            else:
                print("not present")
        else:
            if not self.lchild :
                temp = self.rchild
                self = None
                return temp
            if not self.rchild :
                temp = self.lchild
                self = None
                return temp
            node = self.rchild
            while node.lchild :
                node = node.lchild
            self.key = node.key
            self.rchild = self.rchild.delete(node.key)
        return self
    
    def maxNode(self):
        node=self
        while node.rchild:
            node= node.rchild
        print("maxval is ", node.key)
        

    def minNode(self):
        node=self
        while node.lchild:
            node= node.lchild
        print("minval is ", node.key)


a1=BST(7)
a1.insert(5)
a1.insert(4)
a1.insert(6)
a1.insert(20)
a1.insert(9)
a1.insert(3)
a1.maxNode()
a1.minNode()
a1.delete(20)
a1.preorder()

File: binarySearchTree/bst.cpp
#include <iostream>
using namespace std;


class BST{
private:
    int key;
    BST* left;
    BST* right;
public:
    BST(int data){
        this->key=data;
        this->left=nullptr;
        this->right=nullptr;
    }

    void insert(int data){
        if (data < this->key){
            if (this->left!=nullptr){
                this->left->insert(data); 
            }
            else{
                this->left=new BST(data);
            }
        }else{
            if (this->right!=nullptr){
                this->right->insert(data); 
            }
            else{
                this->right=new BST(data);
            }

        }

    }
    void preOrder(){
        cout<<this->key<<endl;
        if (this->left!=nullptr)
            this->left->preOrder();
        if (this->right!=nullptr)
            this->right->preOrder();
    }

    void inOrder(){
        if (this->left!=nullptr)
            this->left->preOrder();
        cout<<this->key<<endl;
        if (this->right!=nullptr)
            this->right->preOrder();

    }
    void postOrder(){
        if (this->left!=nullptr)
            this->left->postOrder();
        if (this->right!=nullptr)
            this->right->postOrder();
        cout<<this->key<<endl;
    }
    bool search(int data){
        if (this->key==data) return true;
        if (data > this->key){
            if (this->left!=nullptr){
                return this->left->search(data);
            }
            return false;
        }
        else{
            if (this->right!=nullptr){
                return this->right->search(data);
            }
            return false;
        }
    }
    BST* del(int data){
        if (data < this->key){
            if (this->left!=nullptr){
                this->left = this->left->del(data);
            }else{
                cout<<"node not present"<<endl;
                return this;
            }
        }
        else if (data > this->key){
            if (this->right!=nullptr){
                this->right = this->right->del(data);
            }else{
                cout<<"node not present"<<endl;
                return this;
            }   
        }
        else{
            if (this->left!=nullptr){
                BST* temp = this->right;
                delete(this);
                return temp;
            }
            if (this->right!=nullptr){
                BST* temp = this->left;
                delete(this);
                return temp;
            }
            BST* node = this->right;
            while (node->left!=nullptr){
                node=node->left;    
            }
            this->key = node->key;
            this->right = this->right->del(node->key);
    
        }
        return this;
    }

    int minNode(){
        BST* node = this;
        while (node->left != nullptr){
            node = node->left;
        }
        return node->key;
    }
     int maxNode(){
        BST* node = this;
        while (node->right != nullptr){
            node = node->right;
        }
        return node->key;
    }
};

int main(){
    BST* b=new BST(5);
    b->insert(6);
    b->insert(7);
    b->insert(10);
    b->insert(2);
    b->insert(5);
    b->insert(8);
    b->preOrder();
    cout<<boolalpha<<b->search(5)<<endl;
    b->del(10);
    b->postOrder();
    return 0;
}